src/Quasar/FS/Mount.purs

```diff
@@ -19,35 +19,129 @@ module Quasar.FS.Mount where
+  | Couchbase (f DirPath)
+  | MarkLogic (f DirPath)
+  | SparkHDFS (f DirPath)
+  | SparkLocal (f DirPath)
+  | Mimir (f DirPath)
+  | Unknown String (f AnyPath)
 
-derive instance eqMount ∷ Eq Mount
-derive instance ordMount ∷ Ord Mount
-derive instance genericMount ∷ Generic Mount _
-instance showMount ∷ Show Mount where show = genericShow
+type Mount = MountF Identity
+type MountType = MountF (Const Unit)
+
+
+instance eqMount ∷ (Eq1 f) => Eq (MountF f) where
+  eq inX inY =  case inX of
+    View x
+      | View y <- inY -> eq1 x y
+      | otherwise -> false
+    Module x
+      | Module y <- inY -> eq1 x y
+      | otherwise -> false
+    MongoDB x
+      | MongoDB y <- inY -> eq1 x y
+      | otherwise -> false
+    Couchbase x
+      | Couchbase y <- inY -> eq1 x y
+      | otherwise -> false
+    MarkLogic x
+      | MarkLogic y <- inY -> eq1 x y
+      | otherwise -> false
+    SparkHDFS x
+      | SparkHDFS y <- inY -> eq1 x y
+      | otherwise -> false
+    SparkLocal x
+      | SparkLocal y <- inY -> eq1 x y
+      | otherwise -> false
+    Mimir x
+      | Mimir y <- inY -> eq1 x y
+      | otherwise -> false
+    Unknown xName x
+      | Unknown yName y <- inY -> eq1 x y && eq xName yName
+      | otherwise -> false
+
+instance showMount ∷ (Show (f String), Functor f) => Show (MountF f) where
+  show = 
+    let 
+      show' :: forall a. Show a => f a -> String
+      show' = map show >>> show
+    in case _ of
+      View p -> "(View " <> show' p <> ")"
+      Module p -> "(Module " <> show' p <> ")"
+      MongoDB p -> "(MongoDB " <> show' p <> ")"
+      Couchbase p -> "(Couchbase " <> show' p <> ")"
+      MarkLogic p -> "(MarkLogic " <> show' p <> ")"
+      SparkHDFS p -> "(SparkHDFS " <> show' p <> ")"
+      SparkLocal p -> "(SparkLocal " <> show' p <> ")"
+      Mimir p -> "(Mimir " <> show' p <> ")"
+      Unknown n p -> "(Unknown " <> n <> " " <> show' p <> ")"
 
 -- | Attempts to decode a mount listing value from Quasar's filesystem metadata,
 -- | for a mount in the specified parent directory.
-fromJSON ∷ DirPath → Json → Either String Mount
+fromJSON ∷ DirPath → Json → Either String (Mount)
 fromJSON parent = decodeJson >=> \obj → do
   mount ← obj .? "mount"
   typ ← obj .? "type"
   name ← obj .? "name"
-  case typ of
-    "file" → Right $ Mount (MT.fromName mount) (Right (parent </> file name))
-    "directory" → Right $ Mount (MT.fromName mount) (Left (parent </> dir name))
-    _ → Left $ "Unknown type '" <> typ <> "' for mount"
+  let 
+    err :: forall a. Either String a
+    err = Left $ "Unexpected type '" <> typ <> "' for mount '" <> mount <> "'"
+    asFile = parent </> file name
+    asDir = parent </> dir name
+    onFile = if typ == "file" then Right $ Identity asFile else err
+    onDir = if typ == "directory" then Right $ Identity asDir else err
+    onAnyPath = case typ of 
+      "file" → Right $ Identity $ Right asFile
+      "directory" → Right $ Identity $ Left asDir
+      _ → err
```

Comment:
Let's move some of this stuff into a `where` and make it accept arguments so it doesn't compute everything here regardless of whether it is used or not.

Reply:
is current version fine?